### PR TITLE
Added VAS support for magento 2.x

### PR DESCRIPTION
**Pull request by ChetanGN**

### CODE REVIEW

File: Controller/Payment/Order.php [hunk 1]
```diff
@@ -25,7 +50,12 @@ public function __construct(
         \Magento\Checkout\Model\Session $checkoutSession,
         \Razorpay\Magento\Model\CheckoutFactory $checkoutFactory,
         \Razorpay\Magento\Model\Config $config,
-        \Magento\Catalog\Model\Session $catalogSession
+        \Magento\Catalog\Model\Session $catalogSession,
+        \Magento\Quote\Model\QuoteRepository $quoteRepository,
+        \Magento\Sales\Api\Data\OrderInterface $order,
+        \Magento\Quote\Model\QuoteManagement $quoteManagement,
+        \Magento\Store\Model\StoreManagerInterface $storeManagement,
+        \Magento\Customer\Api\CustomerRepositoryInterface $customerRepository
```
Comment (ramth05) [inline Controller/Payment/Order.php]: @ChetanGN , can you please just try use `Magento\Quote\Api\CartManagementInterface` to place the order this much code not required all the details already present in quote in active session. No need to fetch anything or form anything.

**Review by ramth05 — CHANGES_REQUESTED** [review]

**Comment by ChetanGN** [comment]
> Why email in query params?

Basically in both the case normal and hosted checkout, guest user email address will be updated to quote once the payment is done. For the normal checkout after payment there is one ajax call payment-information in which the email is present ok. I think using that it will update it in quote. But in case of hosted once the payment is done it will call our order controller where I am not able to get the user entered mail address from anywhere. So I thought will pass it in query parameter.

**Comment by ramth05** [comment]
> > Why email in query params?
> 
> Basically in both the case normal and hosted checkout, guest user email address will be updated to quote once the payment is done. For the normal checkout after payment there is one ajax call payment-information in which the email is present ok. I think using that it will update it in quote. But in case of hosted once the payment is done it will call our order controller where I am not able to get the user entered mail address from anywhere. So I thought will pass it in query parameter.

But why query params? you can use session

**Review by ramth05 — DISMISSED** [review]

File: Controller/Payment/Order.php [hunk 2]
```diff
@@ -118,83 +137,136 @@ public function execute()
...
+            $response = $api->request->request("GET", "preferences");
         }
-        catch(\Exception $e)
+        catch (Exception $e)
         {
-            $responseContent = [
-                'message'   => $e->getMessage(),
-                'parameters' => []
-            ];
+            echo 'Magento Error : ' . $e->getMessage();
```
Comment (ramth05) [inline Controller/Payment/Order.php:255]: This Should be API exception/error

File: Controller/Payment/Order.php [hunk 3]
```diff
@@ -118,83 +137,136 @@ public function execute()
...
         }
 
-        //set the chache for race with webhook
-        $this->cache->save("started", "quote_Front_processing_$receipt_id", ["razorpay"], 30);
```
Comment (ramth05) [inline Controller/Payment/Order.php:188]: Why this removed?

File: view/frontend/web/js/view/payment/method-renderer/razorpay-method.js [hunk 4]
```diff
@@ -153,6 +161,85 @@ define(
...
+
+                          form.appendChild(self._createHiddenInput(prepareKey, value));
+                        }
+                      }
+                    }
+                }
+              visitNestedOption(options);
+            },
+
+            _createHiddenInput: function(key, value) {
```
Comment (ramth05) [inline view/frontend/web/js/view/payment/method-renderer/razorpay-method.js]: Please remove "_"

File: view/frontend/web/js/view/payment/method-renderer/razorpay-method.js [hunk 5]
```diff
@@ -153,6 +161,85 @@ define(
                 });
             },
 
+            _prepareHiddenFormFields: function (options, form) {
```
Comment (ramth05) [inline view/frontend/web/js/view/payment/method-renderer/razorpay-method.js]: remove "-"

File: view/frontend/web/js/view/payment/method-renderer/razorpay-method.js [hunk 6]
```diff
@@ -153,6 +161,85 @@ define(
...
+                var self = this;
+
+                function visitNestedOption(options, parentKey) {
```
Comment (ramth05) [inline view/frontend/web/js/view/payment/method-renderer/razorpay-method.js:167]: Can we rename this function with "ceateInputFieldsFromOptions"

Comment (ChetanGN) [reply view/frontend/web/js/view/payment/method-renderer/razorpay-method.js:167]: Sure.

File: view/frontend/web/js/view/payment/method-renderer/razorpay-method.js [hunk 7]
```diff
@@ -153,6 +161,85 @@ define(
...
+                }
+                const options = JSON.parse(JSON.stringify(opts));
+
+                var form = document.createElement('form'),
+                    method = 'POST',
+                    input,
+                    key;
+
+                form.method = method;
+                form.action = 'https://api.razorpay.com/v1/checkout/embedded';
```
Comment (ramth05) [inline view/frontend/web/js/view/payment/method-renderer/razorpay-method.js]: Use the base url from API.

File: view/frontend/web/js/view/payment/method-renderer/razorpay-method.js [hunk 8]
```diff
@@ -153,6 +161,85 @@ define(
...
+
+                self._prepareHiddenFormFields(options, form);
```
Comment (ramth05) [inline view/frontend/web/js/view/payment/method-renderer/razorpay-method.js]: what is we directly call "visitNestedOption" from here.

File: Controller/Payment/Webhook.php [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<?php
```
Comment (ramth05) [inline Controller/Payment/Webhook.php]: What changes in this file? if not remove from this PR

File: Controller/Payment/Order.php [hunk 10]
```diff
@@ -2,6 +2,7 @@
 
 namespace Razorpay\Magento\Controller\Payment;
 
+use Razorpay\Api\Api;
```
Comment (ramth05) [inline Controller/Payment/Order.php:5]: Why this referenced?

**Review by ramth05 — CHANGES_REQUESTED** [review]
Please apply the changes.

File: view/frontend/web/js/view/payment/method-renderer/razorpay-method.js (hunk 8 above)
Comment (ChetanGN) [reply view/frontend/web/js/view/payment/method-renderer/razorpay-method.js]: As we need to process object fields also here (prefill, notes) I had done like this.

File: Controller/Payment/Webhook.php (hunk 9 above)
Comment (ChetanGN) [reply Controller/Payment/Webhook.php]: Previously the email already existing issue was fixed in this PR, so this file gets added to this PR.

File: Controller/Payment/Order.php (hunk 10 above)
Comment (ChetanGN) [reply Controller/Payment/Order.php:5]: To call preference API,
```
$api = new Api($this->config->getKeyId(),"");
$response = $api->request->request("GET", "preferences");
```

File: Controller/Payment/Order.php (hunk 3 above)
Comment (ChetanGN) [reply Controller/Payment/Order.php:188]: fixed.

File: Controller/Payment/Order.php (hunk 2 above)
Comment (ChetanGN) [reply Controller/Payment/Order.php:255]: fixed.

File: Controller/Payment/Order.php (hunk 10 above)
Comment (ramth05) [reply Controller/Payment/Order.php:5]: But Order create is already using API reference, before as well.

File: Controller/Payment/Order.php [hunk 11]
```diff
@@ -12,14 +13,23 @@ class Order extends \Razorpay\Magento\Controller\BaseController
     protected $checkoutSession;
 
     protected $_currency = PaymentMethod::CURRENCY;
+
+    protected $cartManagement;
+
+    /**
+     * @var \Magento\Sales\Api\Data\OrderInterface
+     */
+    protected $order;
```
Comment (ramth05) [inline Controller/Payment/Order.php]: is this variable is in use?

File: Controller/Payment/Order.php [hunk 12]
```diff
@@ -12,14 +13,23 @@ class Order extends \Razorpay\Magento\Controller\BaseController
...
+
+    const STATUS_APPROVED = 'APPROVED';
```
Comment (ramth05) [inline Controller/Payment/Order.php]: is this variable is in use?

**Review by ramth05 — APPROVED** [review]
LGTM

**Pull request merged**